crates/types/src/config/log_server.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::num::NonZeroUsize;
+use std::path::PathBuf;
+
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+use restate_serde_util::NonZeroByteCount;
+use tracing::warn;
+
+use super::{CommonOptions, RocksDbOptions, RocksDbOptionsBuilder};
+
+/// # Log server options
+///
+/// Configuration is only used on nodes running with `log-server` role.
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, derive_builder::Builder)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "schemars", schemars(rename = "LogServer", default))]
+#[serde(rename_all = "kebab-case")]
+#[builder(default)]
+pub struct LogServerOptions {
+    #[serde(flatten)]
+    pub rocksdb: RocksDbOptions,
+
+    /// The memory budget for rocksdb memtables in bytes
+    ///
+    /// If this value is set, it overrides the ratio defined in `rocksdb-memory-ratio`.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde_as(as = "Option<NonZeroByteCount>")]
+    #[cfg_attr(feature = "schemars", schemars(with = "Option<NonZeroByteCount>"))]
+    rocksdb_memory_budget: Option<NonZeroUsize>,
+
+    /// The memory budget for rocksdb memtables as ratio
+    ///
+    /// This defines the total memory for rocksdb as a ratio of all memory available to the
+    /// log-server.
+    ///
+    /// (See `rocksdb-total-memtables-ratio` in common).
+    rocksdb_memory_ratio: f32,
+
+    /// Disable fsync of WAL on every batch
+    rocksdb_disable_wal_fsync: bool,
+
+    /// Whether to perform commits in background IO thread pools eagerly or not
+    #[cfg_attr(feature = "schemars", schemars(skip))]
+    #[serde(skip_serializing_if = "std::ops::Not::not", default)]
+    pub always_commit_in_background: bool,
+
+    /// Trigger a commit when the batch size exceeds this threshold.
+    ///
+    /// Set to 0 or 1 to commit the write batch on every command.
+    pub writer_batch_commit_count: usize,
+
+    /// The number of messages that can queue up on input network stream while request processor is busy.
+    pub incoming_network_queue_length: NonZeroUsize,
+}
+
+impl LogServerOptions {
+    pub fn apply_common(&mut self, common: &CommonOptions) {
+        self.rocksdb.apply_common(&common.rocksdb);
+        if self.rocksdb_memory_budget.is_none() {
+            self.rocksdb_memory_budget = Some(
+                // 1MB minimum
+                NonZeroUsize::new(
+                    (common.rocksdb_safe_total_memtables_size() as f64
+                        * self.rocksdb_memory_ratio as f64)
```

Comment:
Is `common.rocksdb_safe_total_memtables_size()` shared between `MetadataStore`, `PartitionStore` and `LogServer`? If it is shared, we might go beyond our memory limit.

Reply:
Yes. We might if ratios add to more than 1, but default should be _sane enough_ (0.01+0.49+0.5)